### PR TITLE
Dropdown menu

### PR DESCRIPTION
**Pull request by a-saran**

### CODE REVIEW

**Comment by Kameshwaran** [comment]
@a-saran Can you rebase this branch over dev and push back?

File: docs/pages/components/DropdownMenu.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      { label: 'Menu Item Four', value: 'D0' },
+      { label: 'Menu Item Five', value: 'E0' },
+      { label: 'Menu Item Six', value: 'F0' },
+      { label: 'Menu Item Seven', value: 'G0' },
+    ]} 
+    onSelect={(value, event) => console.log(value)}
+  >
+    Trigger &#9662;
+  </DropdownMenu>
+```
```
Comment (Kameshwaran) [inline docs/pages/components/DropdownMenu.md]: - [ ] This is not the right way to format editable code. Please check other components doc.

File: docs/theme.js [hunk 2]
```diff
@@ -112,6 +112,29 @@ export default {
...
       borderColor: 'border',
       borderRadius: 2,
-    }
-  }
+    },
+  },
+  dropdownMenu: {
+    primary: {
+      dropdownTrigger: {
+        border: '1px solid #ddd',
```
Comment (Kameshwaran) [inline docs/theme.js]: - [ ] Don't hard code the `borderWidth` and `borderColor`. Those value should be referred from theme.

File: src/components/dropdownMenu.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+import { Box, css } from 'theme-ui';
+import { InlineBlock } from './layout';
+import { PropTypes } from 'prop-types';
+import {
+  applyVariation,
+  UP_ARROW,
+  DOWN_ARROW,
+  ENTER_KEY,
+  TAB_KEY,
+} from './../utils/getStyles'; // TODO: move to separate file after pills merged
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: ??

Comment (Kameshwaran) [reply src/components/dropdownMenu.js]: pills component merged.

File: src/components/dropdownMenu.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+const DropDown = styled(InlineBlock)(
+  ({ theme, variant }) => applyVariation(theme, `${variant}.dropdownTrigger`, 'dropdownMenu')
+);
+
+const MenusContainer = styled(Box)`
+  ${({ theme, displayTop }) => css({
+    position: 'absolute',
+    width: '200px',
+    background: 'white',
+    padding: '10px 0',
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: - [ ] Padding hardcoded?

File: src/components/dropdownMenu.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    borderStyle: 'solid',
+    borderWidth: '1px',
+    borderColor: 'borderGray',
+    borderRadius: '6px',
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: - [ ] borderWidth and borderRadius hardcoded

File: src/components/dropdownMenu.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    bottom: displayTop,
+  })(theme)}
+  ${({theme, variant}) => applyVariation(theme, `${variant}.dropdownContainer`, 'dropdownMenu')}
+`;
+
+const Menu = styled(Box)`
+  ${({ theme }) => css({
+    position: 'relative',
+    padding: '5px 10px',
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: Again padding hardcoded

File: src/components/dropdownMenu.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+const DropdownContainer = styled(InlineBlock)({
+  position: 'relative',
+});
+
+const DropDown = styled(InlineBlock)(
+  ({ theme, variant }) => applyVariation(theme, `${variant}.dropdownTrigger`, 'dropdownMenu')
+);
+
+const MenusContainer = styled(Box)`
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: Why are we using box here?

why dont we use `Absolute`

Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: - [ ] MenusContainer Or OptionsContainer
Which one is better

File: src/components/dropdownMenu.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  ${({ theme, displayTop }) => css({
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: - [ ] Can we name this `displayTop` prop better

File: src/components/dropdownMenu.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+  ${({ theme }) => css({
+    position: 'relative',
+    padding: '5px 10px',
+    '&:hover': {
+      background: 'borderGray',
+    },
+  })(theme)}
+  ${({ theme, variant }) => applyVariation(theme, `${variant}.dropdownOption`, 'dropdownMenu')}
+  ${({ theme, variant, hover }) => hover ? applyVariation(theme, `${variant}.dropdownOption.&:hover`, 'dropdownMenu') : ''}
+  ${({ theme, variant, hover }) => hover ? applyVariation(theme, `${variant}.dropdownOption.&:focus`, 'dropdownMenu') : ''}
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js:51]: - [x] Not sure why we need this

File: src/components/dropdownMenu.js [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        }}
+        key={option.value}
+        onMouseOver={() => setKeySelected(index)}
+        hover={keySelected === index}
+        variant={variant}
+      >
+        {option.label}
+      </Menu>
+    ))
+    : noOptionMessage;
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: - [x] All these functions can be memoized using `useCallback`

**Review by Kameshwaran — COMMENTED** [review]
@a-saran Please look into the comments

File: src/components/dropdownMenu.js (hunk 9 above)
Comment (a-saran) [reply src/components/dropdownMenu.js:51]: In some case the mouse-over style and keyboard focus style can be different, initially hover style is applied when the focus (keyboard) style is not given

File: docs/pages/components/DropdownMenu.md (hunk 1 above)
Comment (a-saran) [reply docs/pages/components/DropdownMenu.md]: do you mean trigger text?

Comment (Kameshwaran) [reply docs/pages/components/DropdownMenu.md]: No. I meant, the editable code. The code formatting should be something like this
Screenshot 2020-04-01 at 8 23 42 AM

File: src/components/dropdownMenu.js [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        onClick={() => toggleDropdown(!showDropdownMenu)}
+        id={id}
+        tabIndex='0'
+        onKeyUp={(e) => e.keyCode === ENTER_KEY ? handleKeyboardEvent(e) : null}
+        ref={dropdownButton}
+      >
+        {children}
+      </DropDown>
+      {showDropdownMenu && (
+        <OptionsContainer
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js:177]: - [x] OptionsContainer should have z-index

File: src/components/dropdownMenu.js [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    '&:hover': {
+      background: 'borderGray',
+    },
+  })(theme)}
+  ${({ theme, variant }) => applyVariation(theme, `${variant}.dropdownOption`, 'dropdownMenu')}
+  ${({ theme, variant, hover }) => hover ? applyVariation(theme, `${variant}.dropdownOption.&:hover`, 'dropdownMenu') : ''}
+  ${({ theme, variant, hover }) => hover ? applyVariation(theme, `${variant}.dropdownOption.&:focus`, 'dropdownMenu') : ''}
+`;
+
+const DropdownMenu = ({
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js:54]: - [ ] Is this a menu component or Input component?

Comment (Kameshwaran) [reply src/components/dropdownMenu.js:54]: - [ ] I guess you developed a menu component whereas what we need is an input.

Check the issue and the reference mentioned
https://github.com/Codebrahma/bricks/issues/14

File: src/components/dropdownMenu.js [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+};
+
+DropdownMenu.defaultProps = {
+  id: '',
+  onSelect: null,
+  variant: 'primary',
+  noOptionMessage: 'No options found',
+  scrollableContainer: null,
+  options: null,
+};
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js:212]: - [ ] If this is an input, it should accept placeholder props.

File: src/components/dropdownMenu.js [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+const DropdownMenu = ({
+  id,
+  variant,
+  children,
+  options,
+  onSelect,
+  noOptionMessage,
+  scrollableContainer,
+}) => {
+  const [showDropdownMenu, setShowDropdownMenu] = useState(false);
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: - [ ] isMenuVisible or isOptionsVisible is a better name

File: src/components/dropdownMenu.js [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    ))
+    : noOptionMessage;
+
+  return (
+    <DropdownContainer ref={componentRef}>
+      <DropDown
+        variant={variant}
+        onClick={() => toggleDropdown(!showDropdownMenu)}
+        id={id}
+        tabIndex='0'
```
Comment (Kameshwaran) [inline src/components/dropdownMenu.js]: - [ ] any reason for hardcoding tabIndex? What if the developer wants to override the tabIndex

File: docs/src/pages/dropdownMenu.md [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    { label: 'Menu Item Four', value: 'D0' },
+    { label: 'Menu Item Five', value: 'E0' },
+    { label: 'Menu Item Six', value: 'F0' },
+    { label: 'Menu Item Seven', value: 'G0' },
+  ]}
+  onSelect={value => alert(JSON.stringify(value))}
+>
+  Trigger &#9662;
+</DropdownMenu>
+```
```
Comment (Kameshwaran) [inline docs/src/pages/dropdownMenu.md:22]: @a-saran I've deleted a component you had written for the demo. No separate component is needed unless you need some state / ref variables.

**Review by Kameshwaran — COMMENTED** [review]
@a-saran

**Review by Kameshwaran — APPROVED** [review]

**Pull request merged**